nx/lib/nx.ex

```diff
@@ -7784,6 +7784,54 @@ defmodule Nx do
     impl!(tensor).take_along_axis(%{tensor | shape: shape}, tensor, indices, axis)
   end
 
+  @doc """
```

Comment:
This looks good! I'm not too sure if this is an API we want to expose just yet, but I can see this being useful for a lot of use-cases.

@josevalim WDYT?

Reply:
This could be the underlying function for:

```elixir
t = Nx.tensor([[1, 2], [3, 4]])
i = Nx.tensor([[0, 1], [0, 0], [1, 1]])

t[i] == Nx.tensor([2, 1, 4])
```

Reply:
Yeah, if you change:

```elixir
def fetch(tensor, %Nx.Tensor{} = index), do: {:ok, fetch_axes(tensor, [{0, index}])}
```

to 

```elixir
def fetch(%Nx.Tensor{} = tensor,%Nx.Tensor{} = index), do: {:ok, Nx.gather(tensor, index)}
```

We can have the behaviour as I described above. Of course, with arbitrary shapes in the output:

```elixir
iex(1)> t = Nx.iota({2, 3, 4})
iex(10)> t[Nx.tensor([[[[[0, 0, 0], [0, 0, 1], [0, 1, 0], [1, 2, 3]]]]])]
#Nx.Tensor<
  s64[1][1][1][4]
  [
    [
      [
        [0, 1, 4, 23]
      ]
    ]
  ]
>
```